Blocks/Persistence/persistence.h

```diff
@@ -132,17 +134,42 @@ class Logic {
   }
 
  protected:
-  size_t DoPublish(const E& entry) {
+  // Deliberately keep these two signatures and not one with `std::forward<>` to ensure the type is right.
+  size_t DoPublish(const ENTRY& entry) {
     std::lock_guard<std::mutex> lock(mutex_);
-    persistence_layer_.Publish(entry);
     list_.push_back(entry);
+    persistence_layer_.Publish(entry);
     return list_.size() - 1;
   }
-
-  size_t DoPublish(E&& entry) {
+  size_t DoPublish(ENTRY&& entry) {
     std::lock_guard<std::mutex> lock(mutex_);
-    persistence_layer_.Publish(static_cast<const E&>(entry));
     list_.push_back(std::move(entry));
+    persistence_layer_.Publish(static_cast<const ENTRY&>(list_.back()));
+    return list_.size() - 1;
+  }
+
+  template <typename DERIVED_E>
+  size_t DoPublishDerived(const DERIVED_E& entry) {
+    static_assert(bricks::can_be_stored_in_unique_ptr<ENTRY, DERIVED_E>::value, "");
+    std::lock_guard<std::mutex> lock(mutex_);
+
+    // `std::unique_ptr<DERIVED>` can be implicitly converted into `std::unique_ptr<BASE>`.
+    // This requires the destructor of `BASE` to be virtual, which is the case for Current and Yoda.
```

Comment:
Just noticed another place where you used `BASE` instead of `ENTRY` :) And it really sounds more natural for the polymorphic case. Maybe we ought to use smth like `BASE_ENTRY` to define a base type template parameter?